// smithers-orchestrator/dist/src/utils/structured-output/zod-converter.js
// Zod to JSON Schema Conversion
// Converts Zod schemas to JSON schema representation for prompts
/**
 * Convert a Zod schema to a JSON schema representation for prompts.
 * Uses Zod 4's built-in toJSONSchema() method.
 */
export function zodToJsonSchema(schema) {
    // Zod 4 has a built-in toJSONSchema method
    if (typeof schema.toJSONSchema === 'function') {
        const jsonSchema = schema.toJSONSchema();
        // Remove $schema to keep it cleaner for prompts
        const { $schema: _$schema, ...rest } = jsonSchema;
        return rest;
    }
    // Fallback for older Zod versions or edge cases
    return convertZodType(schema);
}
export function convertZodType(schema) {
    // Zod 4 uses schema._def.type or schema.type directly
    const def = schema._def || schema.def || {};
    const type = def.type || schema.type;
    // Handle object type
    if (type === 'object') {
        const shape = typeof def.shape === 'function' ? def.shape() : def.shape || {};
        const properties = {};
        const required = [];
        for (const [key, value] of Object.entries(shape)) {
            properties[key] = convertZodType(value);
            // Check if field is optional (Zod 4 uses isOptional method)
            const isOptional = typeof value.isOptional === 'function'
                ? value.isOptional()
                : value._def?.type === 'optional';
            if (!isOptional) {
                required.push(key);
            }
        }
        return {
            type: 'object',
            properties,
            required: required.length > 0 ? required : undefined,
        };
    }
    // Handle array type
    if (type === 'array') {
        const itemType = def.element || def.type;
        return {
            type: 'array',
            items: itemType ? convertZodType(itemType) : {},
        };
    }
    // Handle string type
    if (type === 'string') {
        const result = { type: 'string' };
        if (schema['minLength'])
            result['minLength'] = schema['minLength'];
        if (schema['maxLength'])
            result['maxLength'] = schema['maxLength'];
        if (schema['format'])
            result['format'] = schema['format'];
        return result;
    }
    // Handle number type
    if (type === 'number') {
        return { type: 'number' };
    }
    // Handle boolean type
    if (type === 'boolean') {
        return { type: 'boolean' };
    }
    // Handle null type
    if (type === 'null') {
        return { type: 'null' };
    }
    // Handle enum type
    if (type === 'enum') {
        return { enum: def.values || [] };
    }
    // Handle literal type
    if (type === 'literal') {
        return { const: def.value };
    }
    // Handle union type
    if (type === 'union') {
        const options = def.options || [];
        return {
            oneOf: options.map((opt) => convertZodType(opt)),
        };
    }
    // Handle optional type
    if (type === 'optional') {
        return convertZodType(def.innerType || def.wrapped);
    }
    // Handle nullable type
    if (type === 'nullable') {
        const inner = convertZodType(def.innerType || def.wrapped);
        return { oneOf: [inner, { type: 'null' }] };
    }
    // Default fallback
    return {};
}
/**
 * Generate a human-readable schema description for prompts
 */
export function schemaToPromptDescription(schema) {
    const jsonSchema = zodToJsonSchema(schema);
    return JSON.stringify(jsonSchema, null, 2);
}
