// smithers-orchestrator/dist/src/tools/createSmithersTool.js
/**
 * Create a SmithersTool from options.
 *
 * When used within the Smithers orchestrator, tools receive full context including
 * database access, agent/execution IDs, and proper logging. When used standalone
 * (e.g., MCP server, direct invocation, testing), a stub context is created with:
 * - Empty db object (tools should handle gracefully)
 * - 'stub' agentId/executionId
 * - process.cwd() and process.env
 * - console.log for logging
 *
 * This intentional fallback enables tools to work in MCP/standalone contexts
 * where full Smithers orchestration isn't available.
 *
 * @param options - Tool configuration including name, schema, and execute function
 * @returns A SmithersTool compatible with both Smithers orchestration and standalone use
 *
 * @example
 * ```ts
 * const myTool = createSmithersTool({
 *   name: 'my-tool',
 *   description: 'Does something useful',
 *   inputSchema: z.object({ value: z.string() }),
 *   execute: async (input, context) => {
 *     context.log(`Processing: ${input.value}`)
 *     return { result: input.value.toUpperCase() }
 *   }
 * })
 * ```
 */
export function createSmithersTool(options) {
    const { name, inputSchema, outputSchema, execute, description } = options;
    return {
        name,
        description,
        inputSchema,
        ...(outputSchema ? { outputSchema } : {}),
        execute: async (input, execOptions) => {
            const smithersContext = execOptions.smithers ??
                execOptions.experimental_context;
            const abortSignal = execOptions['abortSignal'];
            // When no Smithers context provided (MCP/standalone usage), create stub context.
            // This enables tools to work outside full orchestration - tools should handle
            // empty db gracefully for standalone compatibility.
            const context = smithersContext
                ? {
                    ...smithersContext,
                    ...(abortSignal ? { abortSignal } : {}),
                }
                : {
                    db: {},
                    agentId: 'stub',
                    executionId: 'stub',
                    cwd: process.cwd(),
                    env: process.env,
                    log: console.log,
                    ...(abortSignal ? { abortSignal } : {}),
                };
            return execute(input, context);
        },
    };
}
