// smithers-orchestrator/dist/src/tools/registry.js
// Tools registry - built-in and custom tool support
// ============================================================================
// BUILT-IN TOOLS REGISTRY
// ============================================================================
/**
 * Registry of built-in tools available in different CLIs
 */
export const BUILTIN_TOOLS = {
    // Claude Code built-in tools
    Read: { cli: 'claude', builtin: true, description: 'Read file contents' },
    Edit: { cli: 'claude', builtin: true, description: 'Edit file contents' },
    Write: { cli: 'claude', builtin: true, description: 'Write new files' },
    Bash: { cli: 'claude', builtin: true, description: 'Execute shell commands' },
    Glob: { cli: 'claude', builtin: true, description: 'Find files by pattern' },
    Grep: { cli: 'claude', builtin: true, description: 'Search file contents' },
    Task: { cli: 'claude', builtin: true, description: 'Launch subagent tasks' },
    WebFetch: { cli: 'claude', builtin: true, description: 'Fetch web content' },
    WebSearch: { cli: 'claude', builtin: true, description: 'Search the web' },
    TodoWrite: { cli: 'claude', builtin: true, description: 'Manage todo lists' },
    // Smithers-specific tools
    Report: { cli: 'smithers', builtin: true, description: 'Report progress to orchestration' },
};
/**
 * Check if a tool name is a built-in tool
 */
export function isBuiltinTool(name) {
    return name in BUILTIN_TOOLS;
}
/**
 * Get tool information
 */
export function getToolInfo(name) {
    if (isBuiltinTool(name)) {
        return BUILTIN_TOOLS[name];
    }
    return null;
}
// ============================================================================
// TOOL SPECIFICATION HELPERS
// ============================================================================
/**
 * Check if a tool spec is a custom Tool object
 */
export function isCustomTool(spec) {
    return (typeof spec === 'object' &&
        spec !== null &&
        'execute' in spec &&
        typeof spec.execute === 'function');
}
/**
 * Check if a tool spec is a legacy Tool (has inputSchema as plain object, not Zod)
 */
export function isLegacyTool(spec) {
    return (typeof spec === 'object' &&
        spec !== null &&
        'execute' in spec &&
        'inputSchema' in spec &&
        typeof spec.inputSchema?.type === 'string');
}
/**
 * Check if a tool spec is a SmithersTool (Zod-based with AI SDK compatibility)
 */
export function isSmithersTool(spec) {
    return (typeof spec === 'object' &&
        spec !== null &&
        'execute' in spec &&
        'inputSchema' in spec &&
        typeof spec.inputSchema?._def === 'object');
}
/**
 * Check if a tool spec is an MCP Server
 */
export function isMCPServer(spec) {
    return typeof spec === 'object' && spec !== null && 'command' in spec && !('execute' in spec);
}
/**
 * Check if a tool spec is a built-in tool name
 */
export function isToolName(spec) {
    return typeof spec === 'string';
}
/**
 * Parse tool specifications into categorized lists.
 *
 * Categorizes each spec based on its structure:
 * - String → builtinTools (tool names like 'Read', 'Bash')
 * - Object with 'command' but no 'execute' → mcpServers
 * - Object with Zod inputSchema (_def property) → smithersTools
 * - Object with JSON Schema inputSchema (type property) → legacyTools + customTools
 * - Other objects with execute → customTools
 *
 * @param specs - Array of tool specifications to categorize
 * @returns Object with categorized tool arrays
 *
 * @example
 * ```ts
 * const { builtinTools, mcpServers, smithersTools } = parseToolSpecs([
 *   'Read',
 *   'Bash',
 *   { name: 'mcp-server', command: 'npx', args: ['@some/mcp'] },
 *   mySmithersTool
 * ])
 * ```
 */
export function parseToolSpecs(specs) {
    const builtinTools = [];
    const customTools = [];
    const smithersTools = [];
    const legacyTools = [];
    const mcpServers = [];
    for (const spec of specs) {
        if (isToolName(spec)) {
            builtinTools.push(spec);
        }
        else if (isMCPServer(spec)) {
            mcpServers.push(spec);
        }
        else if (isSmithersTool(spec)) {
            smithersTools.push(spec);
        }
        else if (isLegacyTool(spec)) {
            customTools.push(spec);
            legacyTools.push(spec);
        }
        else if (isCustomTool(spec)) {
            customTools.push(spec);
        }
    }
    return { builtinTools, customTools, smithersTools, legacyTools, mcpServers };
}
/**
 * Build CLI tool flags from tool specs.
 *
 * Generates command-line flags for passing tools to Claude CLI.
 * Currently only handles built-in tools via --allowedTools flag.
 *
 * @param specs - Array of tool specifications
 * @returns Array of CLI flag strings (e.g., ['--allowedTools', 'Read,Bash,Glob'])
 *
 * @example
 * ```ts
 * const flags = buildToolFlags(['Read', 'Bash', myCustomTool])
 * // Returns: ['--allowedTools', 'Read,Bash']
 * ```
 */
export function buildToolFlags(specs) {
    const { builtinTools } = parseToolSpecs(specs);
    const flags = [];
    // Add --allowedTools flag for built-in tools
    if (builtinTools.length > 0) {
        flags.push('--allowedTools', builtinTools.join(','));
    }
    return flags;
}
