// smithers-orchestrator/dist/src/components/JJ/Describe.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useRef, useReducer } from 'react';
import { useSmithers } from '../SmithersProvider.js';
import { useMountedState, useExecutionMount } from '../../reconciler/hooks.js';
import { useExecutionContext } from '../ExecutionContext.js';
/**
 * JJ Describe component - auto-generates commit message.
 *
 * React pattern: Uses useRef + forceUpdate for fire-and-forget VCS ops.
 * Registers with Ralph for task tracking.
 */
export function Describe(props) {
    const smithers = useSmithers();
    const execution = useExecutionContext();
    const [, forceUpdate] = useReducer((x) => x + 1, 0);
    const statusRef = useRef('pending');
    const descriptionRef = useRef(null);
    const errorRef = useRef(null);
    const taskIdRef = useRef(null);
    const isMounted = useMountedState();
    const shouldExecute = smithers.executionEnabled && execution.isActive;
    useExecutionMount(shouldExecute, () => {
        ;
        (async () => {
            taskIdRef.current = smithers.db.tasks.start('jj-describe');
            try {
                statusRef.current = 'running';
                forceUpdate();
                const diff = await Bun.$ `jj diff`.text();
                const lines = diff.split('\n').length;
                const generatedDescription = `Changes: ${lines} lines modified`;
                await Bun.$ `jj describe -m ${generatedDescription}`.quiet();
                if (isMounted()) {
                    descriptionRef.current = generatedDescription;
                    statusRef.current = 'complete';
                    forceUpdate();
                }
                await smithers.db.vcs.addReport({
                    type: 'progress',
                    title: 'JJ Describe',
                    content: generatedDescription,
                    data: {
                        useAgent: props.useAgent,
                        template: props.template,
                    },
                });
            }
            catch (err) {
                if (isMounted()) {
                    errorRef.current = err instanceof Error ? err : new Error(String(err));
                    statusRef.current = 'error';
                    forceUpdate();
                }
            }
            finally {
                if (taskIdRef.current) {
                    smithers.db.tasks.complete(taskIdRef.current);
                }
            }
        })();
    }, [props.template, props.useAgent, smithers]);
    return (_jsx("jj-describe", { status: statusRef.current, description: descriptionRef.current, error: errorRef.current?.message, "use-agent": props.useAgent, template: props.template, children: props.children }));
}
