// smithers-orchestrator/dist/src/components/JJ/Snapshot.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useRef, useReducer } from 'react';
import { useSmithers } from '../SmithersProvider.js';
import { jjSnapshot, getJJStatus } from '../../utils/vcs.js';
import { useMountedState, useExecutionMount } from '../../reconciler/hooks.js';
import { useExecutionContext } from '../ExecutionContext.js';
/**
 * JJ Snapshot component - creates a JJ snapshot and logs to database.
 *
 * React pattern: Uses useRef + forceUpdate for fire-and-forget VCS ops.
 * Registers with Ralph for task tracking.
 */
export function Snapshot(props) {
    const smithers = useSmithers();
    const execution = useExecutionContext();
    const [, forceUpdate] = useReducer((x) => x + 1, 0);
    const statusRef = useRef('pending');
    const changeIdRef = useRef(null);
    const errorRef = useRef(null);
    const taskIdRef = useRef(null);
    const isMounted = useMountedState();
    const shouldExecute = smithers.executionEnabled && execution.isActive;
    useExecutionMount(shouldExecute, () => {
        ;
        (async () => {
            taskIdRef.current = smithers.db.tasks.start('jj-snapshot');
            try {
                statusRef.current = 'running';
                forceUpdate();
                const result = await jjSnapshot(props.message);
                if (!isMounted())
                    return;
                changeIdRef.current = result.changeId;
                const fileStatus = await getJJStatus();
                await smithers.db.vcs.logSnapshot({
                    change_id: result.changeId,
                    description: result.description,
                    files_modified: fileStatus.modified,
                    files_added: fileStatus.added,
                    files_deleted: fileStatus.deleted,
                });
                if (isMounted()) {
                    statusRef.current = 'complete';
                    forceUpdate();
                }
            }
            catch (err) {
                if (isMounted()) {
                    errorRef.current = err instanceof Error ? err : new Error(String(err));
                    statusRef.current = 'error';
                    forceUpdate();
                }
            }
            finally {
                if (taskIdRef.current) {
                    smithers.db.tasks.complete(taskIdRef.current);
                }
            }
        })();
    }, [props.message, smithers]);
    return (_jsx("jj-snapshot", { status: statusRef.current, "change-id": changeIdRef.current, error: errorRef.current?.message, message: props.message, children: props.children }));
}
