// smithers-orchestrator/dist/src/reactive-sqlite/hooks/context.js
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * React Context Provider for ReactiveDatabase
 *
 * Provides database access throughout the component tree without prop drilling.
 */
import { createContext, useContext } from 'react';
/**
 * Context for the ReactiveDatabase instance
 */
const DatabaseContext = createContext(null);
/**
 * Provider component that makes a ReactiveDatabase available to child components
 *
 * @example
 * ```tsx
 * const db = new ReactiveDatabase(':memory:')
 *
 * function App() {
 *   return (
 *     <DatabaseProvider db={db}>
 *       <UserList />
 *     </DatabaseProvider>
 *   )
 * }
 * ```
 */
export function DatabaseProvider({ db, children }) {
    return (_jsx(DatabaseContext.Provider, { value: db, children: children }));
}
/**
 * Hook to access the ReactiveDatabase from context
 *
 * @throws Error if used outside of a DatabaseProvider
 *
 * @example
 * ```tsx
 * function UserList() {
 *   const db = useDatabase()
 *   const users = db.query('SELECT * FROM users')
 *   return <ul>{users.map(u => <li key={u.id}>{u.name}</li>)}</ul>
 * }
 * ```
 */
export function useDatabase() {
    const db = useContext(DatabaseContext);
    if (!db) {
        throw new Error('useDatabase must be used within a DatabaseProvider');
    }
    return db;
}
/**
 * Hook to optionally access the ReactiveDatabase from context
 * Returns null if not within a provider (useful for hooks that accept optional db)
 */
export function useDatabaseOptional() {
    return useContext(DatabaseContext);
}
export { DatabaseContext };
