// smithers-orchestrator/dist/src/db/artifacts.js
// Artifact tracking module for Smithers DB
import { uuid, now, parseJson } from './utils.js';
const mapArtifact = (row) => {
    if (!row)
        return null;
    return {
        id: row.id,
        execution_id: row.execution_id,
        agent_id: row.agent_id ?? undefined,
        name: row.name,
        type: row.type,
        file_path: row.file_path,
        git_hash: row.git_hash ?? undefined,
        git_commit: row.git_commit ?? undefined,
        summary: row.summary ?? undefined,
        line_count: row.line_count ?? undefined,
        byte_size: row.byte_size ?? undefined,
        metadata: parseJson(row.metadata, {}),
        created_at: new Date(row.created_at),
    };
};
export function createArtifactsModule(ctx) {
    const { rdb, getCurrentExecutionId } = ctx;
    const artifacts = {
        add: (name, type, filePath, agentId, metadata) => {
            if (rdb.isClosed)
                return uuid();
            const currentExecutionId = getCurrentExecutionId();
            if (!currentExecutionId)
                throw new Error('No active execution');
            const id = uuid();
            rdb.run(`INSERT INTO artifacts (id, execution_id, agent_id, name, type, file_path, metadata, created_at)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?)`, [id, currentExecutionId, agentId ?? null, name, type, filePath, JSON.stringify(metadata ?? {}), now()]);
            return id;
        },
        list: (executionId) => {
            if (rdb.isClosed)
                return [];
            return rdb.query('SELECT * FROM artifacts WHERE execution_id = ? ORDER BY created_at', [executionId])
                .map(mapArtifact)
                .filter((a) => a !== null);
        },
    };
    return artifacts;
}
