// smithers-orchestrator/dist/src/db/phases.js
// Phase tracking module for Smithers DB
import { uuid, now } from './utils.js';
export function createPhasesModule(ctx) {
    const { rdb, getCurrentExecutionId, getCurrentPhaseId, setCurrentPhaseId } = ctx;
    const phases = {
        start: (name, iteration = 0) => {
            if (rdb.isClosed)
                return uuid();
            const currentExecutionId = getCurrentExecutionId();
            if (!currentExecutionId)
                throw new Error('No active execution');
            const id = uuid();
            rdb.run(`INSERT INTO phases (id, execution_id, name, iteration, status, started_at, created_at)
         VALUES (?, ?, ?, ?, 'running', ?, ?)`, [id, currentExecutionId, name, iteration, now(), now()]);
            setCurrentPhaseId(id);
            return id;
        },
        complete: (id) => {
            if (rdb.isClosed)
                return;
            const startRow = rdb.queryOne('SELECT started_at FROM phases WHERE id = ?', [id]);
            const durationMs = startRow ? Date.now() - new Date(startRow.started_at).getTime() : null;
            rdb.run(`UPDATE phases SET status = 'completed', completed_at = ?, duration_ms = ? WHERE id = ?`, [now(), durationMs, id]);
            if (getCurrentPhaseId() === id)
                setCurrentPhaseId(null);
        },
        fail: (id) => {
            if (rdb.isClosed)
                return;
            rdb.run(`UPDATE phases SET status = 'failed', completed_at = ? WHERE id = ?`, [now(), id]);
            if (getCurrentPhaseId() === id)
                setCurrentPhaseId(null);
        },
        current: () => {
            if (rdb.isClosed)
                return null;
            const currentPhaseId = getCurrentPhaseId();
            if (!currentPhaseId)
                return null;
            return rdb.queryOne('SELECT * FROM phases WHERE id = ?', [currentPhaseId]);
        },
        list: (executionId) => {
            if (rdb.isClosed)
                return [];
            return rdb.query('SELECT * FROM phases WHERE execution_id = ? ORDER BY created_at', [executionId]);
        },
    };
    return phases;
}
