// smithers-orchestrator/dist/src/db/render-frames.js
// Render frames module for Smithers DB
// Stores snapshots of the SmithersNode tree for time-travel debugging
import { uuid, now } from './utils.js';
export function createRenderFramesModule(ctx) {
    const { rdb, getCurrentExecutionId } = ctx;
    return {
        store: (treeXml, ralphCount = 0) => {
            if (rdb.isClosed)
                return uuid();
            const executionId = getCurrentExecutionId();
            if (!executionId)
                throw new Error('No active execution');
            const id = uuid();
            const sequenceNumber = rdb.queryOne('SELECT COALESCE(MAX(sequence_number), -1) + 1 as next FROM render_frames WHERE execution_id = ?', [executionId])?.next ?? 0;
            rdb.run(`INSERT INTO render_frames (id, execution_id, sequence_number, tree_xml, ralph_count, created_at)
         VALUES (?, ?, ?, ?, ?, ?)`, [id, executionId, sequenceNumber, treeXml, ralphCount, now()]);
            return id;
        },
        get: (id) => {
            if (rdb.isClosed)
                return null;
            return rdb.queryOne('SELECT * FROM render_frames WHERE id = ?', [id]);
        },
        getBySequence: (sequenceNumber) => {
            if (rdb.isClosed)
                return null;
            const executionId = getCurrentExecutionId();
            if (!executionId)
                return null;
            return rdb.queryOne('SELECT * FROM render_frames WHERE execution_id = ? AND sequence_number = ?', [executionId, sequenceNumber]);
        },
        list: () => {
            if (rdb.isClosed)
                return [];
            const executionId = getCurrentExecutionId();
            if (!executionId)
                return [];
            return rdb.query('SELECT * FROM render_frames WHERE execution_id = ? ORDER BY sequence_number ASC', [executionId]);
        },
        listForExecution: (executionId) => {
            if (rdb.isClosed)
                return [];
            return rdb.query('SELECT * FROM render_frames WHERE execution_id = ? ORDER BY sequence_number ASC', [executionId]);
        },
        latest: () => {
            if (rdb.isClosed)
                return null;
            const executionId = getCurrentExecutionId();
            if (!executionId)
                return null;
            return rdb.queryOne('SELECT * FROM render_frames WHERE execution_id = ? ORDER BY sequence_number DESC LIMIT 1', [executionId]);
        },
        count: () => {
            if (rdb.isClosed)
                return 0;
            const executionId = getCurrentExecutionId();
            if (!executionId)
                return 0;
            return rdb.queryOne('SELECT COUNT(*) as count FROM render_frames WHERE execution_id = ?', [executionId])?.count ?? 0;
        },
        nextSequence: () => {
            if (rdb.isClosed)
                return 0;
            const executionId = getCurrentExecutionId();
            if (!executionId)
                return 0;
            return rdb.queryOne('SELECT COALESCE(MAX(sequence_number), -1) + 1 as next FROM render_frames WHERE execution_id = ?', [executionId])?.next ?? 0;
        }
    };
}
