// smithers-orchestrator/dist/src/db/steps.js
// Step tracking module for Smithers DB
import { uuid, now } from './utils.js';
export function createStepsModule(ctx) {
    const { rdb, getCurrentExecutionId, getCurrentPhaseId, getCurrentStepId, setCurrentStepId } = ctx;
    const steps = {
        start: (name) => {
            if (rdb.isClosed) {
                return uuid();
            }
            const currentExecutionId = getCurrentExecutionId();
            const currentPhaseId = getCurrentPhaseId();
            if (!currentExecutionId)
                throw new Error('No active execution');
            const id = uuid();
            rdb.run(`INSERT INTO steps (id, execution_id, phase_id, name, status, started_at, created_at)
         VALUES (?, ?, ?, ?, 'running', ?, ?)`, [id, currentExecutionId, currentPhaseId, name ?? null, now(), now()]);
            setCurrentStepId(id);
            return id;
        },
        complete: (id, vcsInfo) => {
            if (rdb.isClosed)
                return;
            const startRow = rdb.queryOne('SELECT started_at FROM steps WHERE id = ?', [id]);
            const durationMs = startRow ? Date.now() - new Date(startRow.started_at).getTime() : null;
            rdb.run(`UPDATE steps SET status = 'completed', completed_at = ?, duration_ms = ?, snapshot_before = ?, snapshot_after = ?, commit_created = ? WHERE id = ?`, [now(), durationMs, vcsInfo?.snapshot_before ?? null, vcsInfo?.snapshot_after ?? null, vcsInfo?.commit_created ?? null, id]);
            if (getCurrentStepId() === id)
                setCurrentStepId(null);
        },
        fail: (id) => {
            if (rdb.isClosed)
                return;
            rdb.run(`UPDATE steps SET status = 'failed', completed_at = ? WHERE id = ?`, [now(), id]);
            if (getCurrentStepId() === id)
                setCurrentStepId(null);
        },
        current: () => {
            if (rdb.isClosed)
                return null;
            const currentStepId = getCurrentStepId();
            if (!currentStepId)
                return null;
            return rdb.queryOne('SELECT * FROM steps WHERE id = ?', [currentStepId]);
        },
        list: (phaseId) => {
            if (rdb.isClosed)
                return [];
            return rdb.query('SELECT * FROM steps WHERE phase_id = ? ORDER BY created_at', [phaseId]);
        },
        getByExecution: (executionId) => {
            if (rdb.isClosed)
                return [];
            return rdb.query('SELECT * FROM steps WHERE execution_id = ? ORDER BY created_at', [executionId]);
        },
    };
    return steps;
}
