// smithers-orchestrator/dist/src/db/tools.js
// Tool call tracking module for Smithers DB
import { uuid, now, parseJson } from './utils.js';
// Helper to map row to typed object with JSON parsing
const mapToolCall = (row) => {
    if (!row)
        return null;
    return {
        ...row,
        input: parseJson(row.input, {}),
    };
};
export function createToolsModule(ctx) {
    const { rdb, getCurrentExecutionId } = ctx;
    const tools = {
        start: (agentId, toolName, input) => {
            if (rdb.isClosed)
                return uuid();
            const currentExecutionId = getCurrentExecutionId();
            if (!currentExecutionId)
                throw new Error('No active execution');
            const id = uuid();
            const timestamp = now();
            rdb.transaction(() => {
                rdb.run(`INSERT INTO tool_calls (id, agent_id, execution_id, tool_name, input, status, started_at, created_at)
           VALUES (?, ?, ?, ?, ?, 'running', ?, ?)`, [id, agentId, currentExecutionId, toolName, JSON.stringify(input), timestamp, timestamp]);
                rdb.run('UPDATE executions SET total_tool_calls = total_tool_calls + 1 WHERE id = ?', [currentExecutionId]);
                rdb.run('UPDATE agents SET tool_calls_count = tool_calls_count + 1 WHERE id = ?', [agentId]);
            });
            return id;
        },
        complete: (id, output, summary) => {
            const outputSize = Buffer.byteLength(output, 'utf8');
            const timestamp = now();
            rdb.transaction(() => {
                const startRow = rdb.queryOne('SELECT started_at FROM tool_calls WHERE id = ?', [id]);
                const durationMs = startRow ? Date.now() - new Date(startRow.started_at).getTime() : null;
                if (outputSize < 1024) {
                    rdb.run(`UPDATE tool_calls SET status = 'completed', output_inline = ?, output_summary = ?, output_size_bytes = ?, completed_at = ?, duration_ms = ? WHERE id = ?`, [output, summary ?? null, outputSize, timestamp, durationMs, id]);
                }
                else {
                    rdb.run(`UPDATE tool_calls SET status = 'completed', output_summary = ?, output_size_bytes = ?, completed_at = ?, duration_ms = ? WHERE id = ?`, [summary ?? output.slice(0, 200), outputSize, timestamp, durationMs, id]);
                }
            });
        },
        fail: (id, error) => {
            rdb.run(`UPDATE tool_calls SET status = 'failed', error = ?, completed_at = ? WHERE id = ?`, [error, now(), id]);
        },
        list: (agentId) => {
            return rdb.query('SELECT * FROM tool_calls WHERE agent_id = ? ORDER BY created_at', [agentId])
                .map(mapToolCall)
                .filter((t) => t !== null);
        },
        getOutput: (id) => {
            const row = rdb.queryOne('SELECT output_inline FROM tool_calls WHERE id = ?', [id]);
            return row?.output_inline ?? null;
        },
    };
    return tools;
}
