// smithers-orchestrator/dist/src/rate-limits/providers/openai.js
const MODEL_PRICING = {
    'gpt-4o': { input: 5 / 1000000, output: 15 / 1000000 },
    'gpt-4o-mini': { input: 0.15 / 1000000, output: 0.6 / 1000000 },
};
const DEFAULT_BASE_URL = 'https://api.openai.com/v1';
function parseResetToDate(value) {
    if (!value)
        return new Date();
    const trimmed = value.trim();
    const date = new Date(trimmed);
    if (!Number.isNaN(date.getTime()))
        return date;
    const match = trimmed.match(/^(\d+(?:\.\d+)?)(ms|s|m|h|d)$/);
    if (!match)
        return new Date();
    const amount = Number(match[1]);
    const unit = match[2];
    const multipliers = {
        ms: 1,
        s: 1000,
        m: 60000,
        h: 3600000,
        d: 86400000,
    };
    const offsetMs = amount * (unit ? (multipliers[unit] ?? 1) : 1);
    return new Date(Date.now() + offsetMs);
}
export function createOpenAIClient(config) {
    const baseUrl = config.baseUrl ?? DEFAULT_BASE_URL;
    return {
        provider: 'openai',
        async queryStatus(model) {
            const response = await fetch(`${baseUrl}/chat/completions`, {
                method: 'POST',
                headers: {
                    Authorization: `Bearer ${config.apiKey}`,
                    ...(config.organization ? { 'OpenAI-Organization': config.organization } : {}),
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    model,
                    max_tokens: 1,
                    messages: [{ role: 'user', content: '.' }],
                }),
            });
            return this.parseHeaders(response.headers, model);
        },
        parseHeaders(headers, model) {
            const now = new Date();
            const parseNumber = (value) => (value ? Number(value) : 0);
            const requestsLimit = parseNumber(headers?.get('x-ratelimit-limit-requests') ?? null);
            const requestsRemaining = parseNumber(headers?.get('x-ratelimit-remaining-requests') ?? null);
            const tokensLimit = parseNumber(headers?.get('x-ratelimit-limit-tokens') ?? null);
            const tokensRemaining = parseNumber(headers?.get('x-ratelimit-remaining-tokens') ?? null);
            return {
                provider: 'openai',
                model: model ?? 'unknown',
                requests: {
                    limit: requestsLimit,
                    remaining: requestsRemaining,
                    resetsAt: parseResetToDate(headers?.get('x-ratelimit-reset-requests') ?? null),
                },
                inputTokens: {
                    limit: tokensLimit,
                    remaining: tokensRemaining,
                    resetsAt: parseResetToDate(headers?.get('x-ratelimit-reset-tokens') ?? null),
                },
                outputTokens: {
                    limit: 0,
                    remaining: 0,
                    resetsAt: now,
                },
                lastQueried: now,
                stale: false,
            };
        },
        estimateCost(model, tokens) {
            const pricing = MODEL_PRICING[model] ?? MODEL_PRICING['gpt-4o'];
            const input = tokens.input * pricing.input;
            const output = tokens.output * pricing.output;
            return { input, output, total: input + output };
        },
    };
}
