// smithers-orchestrator/dist/src/components/Ralph.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext } from 'react';
import { useSmithers } from './SmithersProvider.js';
import { PhaseRegistryProvider } from './PhaseRegistry.js';
/**
 * Ralph context - DEPRECATED
 *
 * Use useSmithers() and db.tasks.start()/complete() instead:
 *
 * Before:
 * ```tsx
 * const ralph = useContext(RalphContext)
 * ralph?.registerTask()
 * // ... do work ...
 * ralph?.completeTask()
 * ```
 *
 * After:
 * ```tsx
 * const { db } = useSmithers()
 * const taskId = db.tasks.start('component-type', 'component-name')
 * try {
 *   // ... do work ...
 * } finally {
 *   db.tasks.complete(taskId)
 * }
 * ```
 *
 * For ralphCount, use useRalphCount() hook instead.
 *
 * @deprecated Use useSmithers() and db.tasks instead
 */
export const RalphContext = createContext(undefined);
// Re-export orchestration signals from SmithersProvider
export { createOrchestrationPromise, signalOrchestrationComplete, signalOrchestrationError, signalOrchestrationCompleteByToken, signalOrchestrationErrorByToken, setActiveOrchestrationToken, } from './SmithersProvider.js';
/**
 * Ralph component - DEPRECATED backwards compatibility wrapper
 *
 * The Ralph loop is now managed by SmithersProvider directly, and task
 * tracking is now database-backed via the tasks table.
 *
 * This component exists for:
 * 1. Backwards compatibility with existing code using <Ralph>
 * 2. Providing RalphContext for components using useContext(RalphContext)
 * 3. Rendering the <ralph> custom element for XML serialization
 *
 * New code should use SmithersProvider directly without Ralph:
 * ```tsx
 * <SmithersProvider db={db} executionId={id} maxIterations={10}>
 *   <Orchestration>
 *     <Claude>...</Claude>
 *   </Orchestration>
 * </SmithersProvider>
 * ```
 *
 * Task tracking in components should use db.tasks:
 * ```tsx
 * const { db } = useSmithers()
 * const ralphCount = useRalphCount()
 * const taskId = db.tasks.start('component-type')
 * try { ... } finally { db.tasks.complete(taskId) }
 * ```
 *
 * @deprecated Use SmithersProvider directly with db.tasks for task tracking
 */
export function Ralph(props) {
    // Get context from SmithersProvider
    const smithers = useSmithers();
    // Build RalphContext value from SmithersProvider
    // Note: registerTask/completeTask are deprecated no-ops
    const contextValue = {
        registerTask: smithers.registerTask,
        completeTask: smithers.completeTask,
        ralphCount: smithers.ralphCount,
        db: smithers.reactiveDb,
    };
    // Note: maxIterations, onIteration, onComplete from props are ignored here
    // since they should be passed to SmithersProvider instead.
    // This component is primarily for backwards compatibility.
    return (_jsx(RalphContext.Provider, { value: contextValue, children: _jsx(PhaseRegistryProvider, { children: _jsx("ralph", { iteration: smithers.ralphCount, maxIterations: props.maxIterations ?? smithers.config.maxIterations ?? 100, children: props.children }) }) }));
}
