// smithers-orchestrator/dist/src/hooks/ai-sdk.js
/**
 * Re-exports from @ai-sdk/react (optional dependency).
 *
 * These hooks provide streaming AI capabilities for React applications.
 * Currently commented out due to module resolution issues.
 *
 * @see https://ai-sdk.dev/docs/reference/ai-sdk-react
 */
export {};
// ============================================================================
// CHAT HOOKS
// ============================================================================
// export { useChat, Chat } from '@ai-sdk/react'
// export type { UseChatOptions, UseChatHelpers } from '@ai-sdk/react'
// ============================================================================
// COMPLETION HOOKS
// ============================================================================
// export { useCompletion } from '@ai-sdk/react'
// export type { UseCompletionOptions, UseCompletionHelpers } from '@ai-sdk/react'
// ============================================================================
// OBJECT/STRUCTURED OUTPUT HOOKS
// ============================================================================
// export { experimental_useObject } from '@ai-sdk/react'
// export type {
//   Experimental_UseObjectOptions,
//   Experimental_UseObjectHelpers,
// } from '@ai-sdk/react'
// ============================================================================
// TYPES
// ============================================================================
// export type { UIMessage, CreateUIMessage } from '@ai-sdk/react'
