// smithers-orchestrator/dist/src/tui/utils/format.js
// Shared formatting utilities for TUI components
export function truncate(str, maxLen, ellipsis = '...') {
    if (str.length <= maxLen)
        return str;
    return str.slice(0, maxLen - ellipsis.length) + ellipsis;
}
export function truncateTilde(str, maxLen) {
    if (str.length <= maxLen)
        return str;
    return str.slice(0, maxLen - 1) + '~';
}
export function formatTimestamp(timestamp) {
    const date = new Date(timestamp);
    if (isNaN(date.getTime())) {
        return timestamp;
    }
    return date.toLocaleString();
}
export function formatTime(timestamp) {
    const date = new Date(timestamp);
    if (isNaN(date.getTime())) {
        return '--:--:--';
    }
    return date.toLocaleTimeString('en-US', {
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit',
        hour12: false
    });
}
export function formatValue(value) {
    if (value === null || value === undefined)
        return 'NULL';
    if (typeof value === 'object')
        return JSON.stringify(value).slice(0, 20);
    return String(value);
}
