// smithers-orchestrator/dist/src/tui/components/layout/TabBar.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { TextAttributes } from '@opentui/core';
export function TabBar({ tabs, activeTab }) {
    return (_jsx("box", { style: {
            height: 1,
            width: '100%',
            backgroundColor: '#16161e',
            flexDirection: 'row',
            paddingLeft: 1
        }, children: tabs.map((tab) => {
            const isActive = tab.key === activeTab;
            return (_jsxs("box", { style: {
                    flexDirection: 'row',
                    marginRight: 2
                }, children: [_jsx("text", { content: `${tab.shortcut}:`, style: {
                            fg: '#565f89',
                        } }), _jsx("text", { content: tab.label, style: {
                            fg: isActive ? '#7aa2f7' : '#a9b1d6',
                            backgroundColor: isActive ? '#24283b' : undefined,
                            attributes: isActive ? TextAttributes.BOLD : undefined,
                            paddingLeft: 1,
                            paddingRight: 1
                        } })] }, tab.key));
        }) }));
}
