// smithers-orchestrator/dist/src/tui/components/views/ChatInterface.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
// Chat Interface View (F4)
// Claude-powered Q&A about execution state
import { useCallback } from 'react';
import { useKeyboard } from '@opentui/react';
import { useClaudeChat } from '../../hooks/useClaudeChat.js';
import { TextAttributes } from '@opentui/core';
import { useTuiState } from '../../state.js';
export function ChatInterface({ db }) {
    const { messages, isLoading, error, isAvailable, sendMessage, clearHistory } = useClaudeChat(db);
    const [inputValue, setInputValue] = useTuiState('tui:chat:input', '');
    const [isInputFocused, setIsInputFocused] = useTuiState('tui:chat:focus', true);
    const handleSubmit = useCallback(async () => {
        if (!inputValue.trim())
            return;
        const message = inputValue.trim();
        setInputValue('');
        await sendMessage(message);
    }, [inputValue, sendMessage]);
    // Handle keyboard
    useKeyboard((key) => {
        if (key.ctrl && key.name === 'l') {
            clearHistory();
        }
        if (key.name === 'tab') {
            setIsInputFocused(prev => !prev);
        }
    });
    if (!isAvailable) {
        return (_jsxs("box", { style: { flexDirection: 'column' }, children: [_jsx("text", { content: "Claude Chat Unavailable", style: { fg: '#f7768e', attributes: TextAttributes.BOLD, marginBottom: 1 } }), _jsx("text", { content: "Set ANTHROPIC_API_KEY environment variable to enable Claude chat.", style: { fg: '#565f89' } }), _jsx("text", { content: "The TUI will show execution data but AI analysis is disabled.", style: { fg: '#414868', marginTop: 1 } })] }));
    }
    return (_jsxs("box", { style: { flexDirection: 'column', width: '100%', height: '100%' }, children: [_jsx("text", { content: "Claude Chat - Ask about your execution", style: { fg: '#7aa2f7', attributes: TextAttributes.BOLD, marginBottom: 1 } }), error && (_jsx("text", { content: `Error: ${error}`, style: { fg: '#f7768e', marginBottom: 1 } })), _jsxs("scrollbox", { focused: !isInputFocused, style: {
                    flexGrow: 1,
                    border: true,
                    padding: 1,
                    backgroundColor: '#16161e',
                    marginBottom: 1
                }, children: [messages.length === 0 ? (_jsxs("box", { style: { flexDirection: 'column' }, children: [_jsx("text", { content: "No messages yet. Try asking:", style: { fg: '#565f89', marginBottom: 1 } }), _jsx("text", { content: '  "What is the current status?"', style: { fg: '#7dcfff' } }), _jsx("text", { content: '  "Show me recent errors"', style: { fg: '#7dcfff' } }), _jsx("text", { content: '  "How many tokens have been used?"', style: { fg: '#7dcfff' } })] })) : (messages.map((msg, index) => (_jsxs("box", { style: { marginBottom: 1 }, children: [_jsx("text", { content: msg.role === 'user' ? 'You: ' : 'Claude: ', style: {
                                    fg: msg.role === 'user' ? '#9ece6a' : '#7aa2f7',
                                    attributes: TextAttributes.BOLD
                                } }), _jsx("text", { content: msg.content, style: { fg: '#c0caf5' } })] }, index)))), isLoading && (_jsx("text", { content: "Claude is thinking...", style: { fg: '#e0af68' } }))] }), _jsx("box", { style: { height: 3, border: true, padding: 1 }, children: _jsx("input", { placeholder: "Ask Claude about the execution...", value: inputValue, focused: isInputFocused, onInput: setInputValue, onSubmit: handleSubmit, style: {
                        width: '100%',
                        focusedBackgroundColor: '#24283b'
                    } }) }), _jsx("text", { content: "Enter to send, Ctrl+L to clear history, Tab to switch focus", style: { fg: '#565f89', marginTop: 1 } })] }));
}
