// smithers-orchestrator/dist/src/tui/components/views/DatabaseExplorer.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
// Database Explorer View (F3)
// Browse SQLite tables and query data
import { useKeyboard } from '@opentui/react';
import { TextAttributes } from '@opentui/core';
import { usePollTableData } from '../../hooks/usePollTableData.js';
import { truncateTilde, formatValue } from '../../utils/format.js';
import { useEffectOnValueChange } from '../../../reconciler/hooks.js';
import { useTuiState } from '../../state.js';
const TABLES = [
    'executions',
    'phases',
    'agents',
    'tool_calls',
    'human_interactions',
    'render_frames',
    'tasks',
    'steps',
    'reports',
    'memories',
    'state',
    'transitions',
    'artifacts',
    'commits',
    'snapshots',
    'reviews'
];
export function DatabaseExplorer({ db, height }) {
    const [selectedTable, setSelectedTable] = useTuiState('tui:db:selectedTable', 0);
    const [rowOffset, setRowOffset] = useTuiState('tui:db:rowOffset', 0);
    const [isTableListFocused, setIsTableListFocused] = useTuiState('tui:db:isTableListFocused', true);
    const tableName = TABLES[selectedTable] ?? 'executions';
    const { columns, data: tableData } = usePollTableData(db, tableName);
    const visibleRowsCount = height - 8;
    useEffectOnValueChange(`${tableData.length}:${visibleRowsCount}`, () => {
        const maxOffset = Math.max(0, tableData.length - visibleRowsCount);
        if (rowOffset > maxOffset) {
            setRowOffset(maxOffset);
        }
    }, [rowOffset, setRowOffset, tableData.length, visibleRowsCount]);
    const handleSelectTable = (index) => {
        setSelectedTable(index);
        setRowOffset(0);
    };
    useKeyboard((key) => {
        if (key.name === 'tab') {
            setIsTableListFocused(!isTableListFocused);
        }
        else if (isTableListFocused) {
            if (key.name === 'j' || key.name === 'down') {
                handleSelectTable(Math.min(selectedTable + 1, TABLES.length - 1));
            }
            else if (key.name === 'k' || key.name === 'up') {
                handleSelectTable(Math.max(selectedTable - 1, 0));
            }
        }
        else {
            if (key.name === 'j' || key.name === 'down') {
                const maxOffset = Math.max(0, tableData.length - visibleRowsCount);
                setRowOffset(prev => Math.min(prev + 1, maxOffset));
            }
            else if (key.name === 'k' || key.name === 'up') {
                setRowOffset(prev => Math.max(prev - 1, 0));
            }
        }
    });
    const visibleRows = tableData.slice(rowOffset, rowOffset + visibleRowsCount);
    return (_jsxs("box", { style: { flexDirection: 'row', width: '100%', height: '100%' }, children: [_jsxs("box", { style: {
                    width: 20,
                    flexDirection: 'column',
                    borderRight: true,
                    paddingRight: 1
                }, children: [_jsx("text", { content: "Tables", style: {
                            fg: '#7aa2f7',
                            attributes: TextAttributes.BOLD,
                            marginBottom: 1
                        } }), _jsx("scrollbox", { focused: isTableListFocused, style: { flexGrow: 1 }, children: TABLES.map((table, index) => (_jsx("text", { content: table, style: {
                                fg: index === selectedTable ? '#7aa2f7' : '#a9b1d6',
                                backgroundColor: index === selectedTable && isTableListFocused ? '#24283b' : undefined,
                                paddingLeft: 1
                            } }, table))) })] }), _jsxs("box", { style: { flexGrow: 1, flexDirection: 'column', paddingLeft: 1 }, children: [_jsxs("box", { style: { flexDirection: 'row', marginBottom: 1, justifyContent: 'space-between' }, children: [_jsx("text", { content: `${tableName} (${tableData.length} rows)`, style: { fg: '#7aa2f7', attributes: TextAttributes.BOLD } }), _jsx("text", { content: "Tab to switch focus, j/k to navigate", style: { fg: '#565f89' } })] }), columns.length > 0 && (_jsxs("box", { style: { flexDirection: 'row', marginBottom: 1 }, children: [columns.slice(0, 5).map((col) => (_jsx("text", { content: truncateTilde(col, 15), style: {
                                    fg: '#bb9af7',
                                    width: 16,
                                    attributes: TextAttributes.BOLD
                                } }, col))), columns.length > 5 && (_jsx("text", { content: `+${columns.length - 5} more`, style: { fg: '#565f89' } }))] })), _jsxs("scrollbox", { focused: !isTableListFocused, style: { flexGrow: 1 }, children: [visibleRows.map((row, index) => (_jsx("box", { style: { flexDirection: 'row' }, children: columns.slice(0, 5).map((col) => (_jsx("text", { content: truncateTilde(formatValue(row[col]), 15), style: {
                                        fg: '#c0caf5',
                                        width: 16
                                    } }, col))) }, index))), tableData.length === 0 && (_jsx("text", { content: "No data", style: { fg: '#565f89' } }))] })] })] }));
}
