// smithers-orchestrator/dist/src/tui/components/views/ReportViewer.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
// Report Viewer View (F6)
// Auto-generated 10-minute summaries
import { useKeyboard } from '@opentui/react';
import { useReportGenerator } from '../../hooks/useReportGenerator.js';
import { TextAttributes } from '@opentui/core';
import { truncate, formatTimestamp } from '../../utils/format.js';
import { getSeverityColor, colors } from '../../utils/colors.js';
import { useEffectOnValueChange } from '../../../reconciler/hooks.js';
import { useTuiState } from '../../state.js';
export function ReportViewer({ db }) {
    const { reports, isGenerating, generateNow } = useReportGenerator(db);
    const [selectedIndex, setSelectedIndex] = useTuiState('tui:reports:selectedIndex', 0);
    useEffectOnValueChange(reports.length, () => {
        const maxIndex = Math.max(0, reports.length - 1);
        if (selectedIndex > maxIndex) {
            setSelectedIndex(maxIndex);
        }
    }, [reports.length, selectedIndex, setSelectedIndex]);
    // Handle keyboard navigation
    useKeyboard((key) => {
        if (key.name === 'j' || key.name === 'down') {
            setSelectedIndex(prev => Math.min(prev + 1, Math.max(0, reports.length - 1)));
        }
        else if (key.name === 'k' || key.name === 'up') {
            setSelectedIndex(prev => Math.max(prev - 1, 0));
        }
        else if (key.name === 'r') {
            generateNow();
        }
    });
    const selectedReport = reports[selectedIndex];
    const hasApiKey = !!process.env['ANTHROPIC_API_KEY'];
    return (_jsxs("box", { style: { flexDirection: 'column', width: '100%', height: '100%' }, children: [_jsxs("box", { style: { flexDirection: 'row', marginBottom: 1, justifyContent: 'space-between' }, children: [_jsx("text", { content: "Auto-Generated Reports", style: { fg: '#7aa2f7', attributes: TextAttributes.BOLD } }), _jsxs("box", { style: { flexDirection: 'row', gap: 2 }, children: [isGenerating && (_jsx("text", { content: "Generating...", style: { fg: '#e0af68' } })), _jsx("text", { content: "[r] Generate Now", style: { fg: '#9ece6a' } })] })] }), !hasApiKey && (_jsx("text", { content: "Note: ANTHROPIC_API_KEY not set - reports show metrics only, no AI analysis", style: { fg: '#e0af68', marginBottom: 1 } })), _jsxs("box", { style: { flexDirection: 'row', height: '100%' }, children: [_jsxs("box", { style: {
                            width: 35,
                            flexDirection: 'column',
                            borderRight: true,
                            paddingRight: 1
                        }, children: [_jsx("text", { content: `Reports (${reports.length})`, style: { fg: '#bb9af7', marginBottom: 1 } }), _jsx("scrollbox", { focused: true, style: { flexGrow: 1 }, children: reports.length === 0 ? (_jsx("text", { content: "No reports yet", style: { fg: '#565f89' } })) : (reports.map((report, index) => (_jsx("box", { style: {
                                        backgroundColor: index === selectedIndex ? '#24283b' : undefined,
                                        paddingLeft: 1
                                    }, children: _jsx("text", { content: truncate(report.title, 30), style: {
                                            fg: getSeverityColor(report.severity),
                                        } }) }, report.id)))) })] }), _jsx("box", { style: { flexGrow: 1, flexDirection: 'column', paddingLeft: 1 }, children: selectedReport ? (_jsxs(_Fragment, { children: [_jsx("text", { content: selectedReport.title, style: {
                                        fg: getSeverityColor(selectedReport.severity),
                                        attributes: TextAttributes.BOLD,
                                        marginBottom: 1
                                    } }), _jsx("text", { content: formatTimestamp(selectedReport.created_at), style: { fg: '#565f89', marginBottom: 1 } }), _jsx("scrollbox", { style: {
                                        flexGrow: 1,
                                        border: true,
                                        padding: 1,
                                        backgroundColor: '#16161e'
                                    }, children: selectedReport.content.split('\n').map((line, index) => (_jsx("text", { content: line, style: { fg: getLineColor(line) } }, index))) })] })) : (_jsxs("box", { style: { flexDirection: 'column' }, children: [_jsx("text", { content: "No report selected", style: { fg: '#565f89', marginBottom: 1 } }), _jsx("text", { content: "Reports are generated every 10 minutes automatically.", style: { fg: '#414868' } }), _jsx("text", { content: "Press [r] to generate one now.", style: { fg: '#414868' } })] })) })] }), _jsx("text", { content: "j/k to navigate, r to generate new report", style: { fg: '#565f89', marginTop: 1 } })] }));
}
function getLineColor(line) {
    if (line.startsWith('##'))
        return colors.purple;
    if (line.startsWith('###'))
        return colors.blue;
    if (line.startsWith('-'))
        return colors.cyan;
    if (line.includes('Error') || line.includes('Failed'))
        return colors.red;
    if (line.includes('Warning'))
        return colors.orange;
    return colors.fg;
}
